### PR TITLE
Second try on Von Mises: Make it JIT compatible

### PR DESCRIPTION
**Pull request by ahmadsalim**
Follow up from #17168 .

### CODE REVIEW

**Review by fritzo — COMMENTED** [review]
Can you confirm that the only change from #17168 (which I already reviewed) are the factoring out of `_rejection_sample()`?

Also nit: I believe flake8 wants another newline before `class VonMises` and one fewer newline before `def sample`.

**Comment by ahmadsalim** [comment]
@fritzo Yes, and also I had to fix the inline `|=` to a normal assignment.

File: torch/distributions/von_mises.py [hunk 1]
```diff
@@ -61,14 +61,15 @@ def _rejection_sample(loc, concentration, proposal_r, shape):
         u = torch.rand((3,) + shape, dtype=loc.dtype, device=loc.device)
         u1, u2, u3 = u.unbind()
         z = torch.cos(math.pi * u1)
-        f = (1 + _proposal_r * z) / (_proposal_r + z)
-        c = concentration * (_proposal_r - f)
+        f = (1 + proposal_r * z) / (proposal_r + z)
+        c = concentration * (proposal_r - f)
         accept = ((c * (2 - c) - u2) > 0) | ((c / u2).log() + 1 - c >= 0)
         if accept.any():
             x[accept] = torch.sign(u3[accept] - 0.5) * torch.acos(f[accept])
```
Comment (fritzo) [inline torch/distributions/von_mises.py]: Note if the in-place op is giving you trouble you can instead
```py
x = torch.where(accept, (u3 - 0.5).sign() * f.acos(), x)
```
I'm don't understand what is causing the test failures.

**Review by fritzo — APPROVED** [review]

**Comment by fritzo** [comment]
LGTM as soon as tests pass.

File: torch/distributions/von_mises.py (hunk 1 above)
Comment (ahmadsalim) [reply torch/distributions/von_mises.py]: Thanks! It looks like the tests failed because `torch.Size` somehow is not supported by TorchScript.

**Review by fritzo — APPROVED** [review]
LGTM. Test failures appear unrelated

<details>

```
/var/lib/jenkins/workspace/scripts/onnx/test.sh: line 57: 26027 Segmentation fault      (core dumped) pytest "${args[@]}" --ignore "$top_dir/test/onnx/test_pytorch_onnx_onnxruntime.py" --ignore "$top_dir/test/onnx/test_custom_ops.py" --ignore "$top_dir/test/onnx/test_models_onnxruntime.py" "${test_paths[@]}"
```

```
Feb 13 03:29:04 ======================================================================
Feb 13 03:29:04 ERROR [300.046s]: test_backward_python_udf_error (__main__.DistAutogradTestWithSpawn)
Feb 13 03:29:04 ----------------------------------------------------------------------
Feb 13 03:29:04 Traceback (most recent call last):
Feb 13 03:29:04   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 175, in wrapper
Feb 13 03:29:04     self._join_processes(fn)
Feb 13 03:29:04   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 256, in _join_processes
Feb 13 03:29:04     self._check_return_codes(elapsed_time)
Feb 13 03:29:04   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 275, in _check_return_codes
Feb 13 03:29:04     raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time))
Feb 13 03:29:04 RuntimeError: Process 0 terminated or timed out after 300.025319814682 seconds
Feb 13 03:29:04 
Feb 13 03:29:04 ----------------------------------------------------------------------
Feb 13 03:29:04 Ran 54 tests in 370.603s
Feb 13 03:29:04 
Feb 13 03:29:04 FAILED (errors=1, skipped=2)
```

</details

**Comment by fritzo** [comment]
@ezyang can you confirm test errors are unrelated?

**Comment by ezyang** [comment]
I agree

**Comment by zhangguanheng66** [comment]
Revert the diff. The failure on the master branch seems relevant.
https://circleci.com/gh/pytorch/pytorch/4495118?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

File: torch/distributions/von_mises.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    # compute large solution
+    y = 3.75 / x
+    large = x - 0.5 * x.log() + _eval_poly(y, _COEF_LARGE[order]).log()
+
+    mask = (x < 3.75)
+    result = large
+    if mask.any():
+        result[mask] = small[mask]
+    return result
```
Comment (fritzo) [inline torch/distributions/von_mises.py]: @ahmadsalim could you try also wrapping this in `@torch.jit.script` and replacing the in-place modification with a `torch.where()`? I think the `mask.any()` conditional may be fooling the jit tracer, but this is just a guess.

**Review by fritzo — COMMENTED** [review]
😞 sorry I missed that error
```
Traceback (most recent call last):
Feb 14 16:04:53   File "test_distributions.py", line 4517, in test_log_prob
Feb 14 16:04:53     message='{}\nExpected:\n{}\nActual:\n{}'.format(Dist.__name__, expected, actual))
Feb 14 16:04:53   File "/opt/python/3.5/lib/python3.5/site-packages/torch/testing/_internal/common_utils.py", line 862, in assertEqual
Feb 14 16:04:53     assertTensorsEqual(x, y)
Feb 14 16:04:53   File "/opt/python/3.5/lib/python3.5/site-packages/torch/testing/_internal/common_utils.py", line 832, in assertTensorsEqual
Feb 14 16:04:53     self.assertLessEqual(max_err, prec, message)
Feb 14 16:04:53 AssertionError: tensor(0.0003) not less than or equal to 1e-05 : VonMises
Feb 14 16:04:53 Expected:
Feb 14 16:04:53 -2.7234730628735986
Feb 14 16:04:53 Actual:
Feb 14 16:04:53 -2.7238097111149546
```
Should we open up a third PR attempt?

**Comment by fritzo** [comment]
I wonder if this is conflicting with https://github.com/pytorch/pytorch/pull/22742 due to random seed churn.

**Comment by zhangguanheng66** [comment]
> 😞 sorry I missed that error
> 
> ```
> Traceback (most recent call last):
> Feb 14 16:04:53   File "test_distributions.py", line 4517, in test_log_prob
> Feb 14 16:04:53     message='{}\nExpected:\n{}\nActual:\n{}'.format(Dist.__name__, expected, actual))
> Feb 14 16:04:53   File "/opt/python/3.5/lib/python3.5/site-packages/torch/testing/_internal/common_utils.py", line 862, in assertEqual
> Feb 14 16:04:53     assertTensorsEqual(x, y)
> Feb 14 16:04:53   File "/opt/python/3.5/lib/python3.5/site-packages/torch/testing/_internal/common_utils.py", line 832, in assertTensorsEqual
> Feb 14 16:04:53     self.assertLessEqual(max_err, prec, message)
> Feb 14 16:04:53 AssertionError: tensor(0.0003) not less than or equal to 1e-05 : VonMises
> Feb 14 16:04:53 Expected:
> Feb 14 16:04:53 -2.7234730628735986
> Feb 14 16:04:53 Actual:
> Feb 14 16:04:53 -2.7238097111149546
> ```
> 
> Should we open up a third PR attempt?

Yeap. I think you should try to re-land the PR. I just reverted it to recover our master branch.

**Comment by ezyang** [comment]
NB: the test config that failed isn't run on PRs. So you'll need to manually enable it; it's pytorch_linux_xenial_py3_5_test. I think if you put `[ci all]` in your commit message that will be good enough; just check the relevant job in CI.

**Comment by ahmadsalim** [comment]
Thanks for the tips! I will try again 😄

**Pull request closed without merging**